### PR TITLE
add TypeBase class and TypeKind.def

### LINKED ISSUE #82: Create `glu::types::TypeBase` (base class for all types) (in AST/Types folder)

### PR DESCRIPTION
**Pull request by badermarouan**
Adds two features from #82

### CODE REVIEW

File: lib/AST/CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,7 @@ add_library (AST SHARED)
 target_include_directories(AST
     PUBLIC
     ${PROJECT_SOURCE_DIR}/include/AST
+    ${PROJECT_SOURCE_DIR}/include/AST/Types
 )
```
Comment (Snowy1803) [inline lib/AST/CMakeLists.txt:7]:
```suggestion
target_include_directories(AST
    PRIVATE
    ${PROJECT_SOURCE_DIR}/include/AST
)
```

File: include/AST/Types/TypeBase.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        EnumTyKind,
+
+        /// Represents an integer type.
+        /// Includes various bit-width integer types (e.g., int32, int64).
+        IntTyKind,
+
+        /// Represents a floating-point type.
+        /// Includes types such as float, double, etc.
+        FloatTyKind,
+
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:34]: .def

File: include/AST/Types/TypeBase.hpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    /// @param size Defines the size of the type in Bytes.
+    /// @param kind Defines the kind of the Type to initialize it dynamicaly
+    ///             with LLVM RTTI
+    TypeBase(std::string name, size_t size, TypeKind kind)
+        : _kind(kind)
+    {
+    }
+
+private:
+    TypeKind const _kind;
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:81]: start of class

File: include/AST/Types/TypeBase.hpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    ///        for LLVM RTTI to dynamicaly define class.
+    /// @param binding Defines whether the type is constant or variable.
+    /// @param name Defines the name of the type as a String.
+    /// @param size Defines the size of the type in Bytes.
+    /// @param kind Defines the kind of the Type to initialize it dynamicaly
+    ///             with LLVM RTTI
+    TypeBase(std::string name, size_t size, TypeKind kind)
+        : _kind(kind)
+    {
+    }
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:78]: no name no size

File: include/AST/Types/TypeBase.hpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        PointerTyKind,
+
+        /// Represents a type alias.
+        /// Used to refer to types through aliases (e.g., typedef in C/C++).
+        TypeAliasTyKind,
+    };
+
+    /// @brief Getter for the kind of the Type.
+    /// @return Returns the size of the type as a TypeKind.
+    inline TypeKind getKind() const
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:63]: useless inline keyword

Comment (LindonAliu) [reply include/AST/Types/TypeBase.hpp:63]: arrete de commenter pour commenter

File: include/AST/Types/TypeBase.hpp (hunk 3 above)
Comment (badermarouan) [reply include/AST/Types/TypeBase.hpp:39]: I need to define TypeKind before

File: include/AST/Types/Typekind.def [hunk 6]
```diff
@@ -0,0 +1,42 @@
+/// Represents a structure type.
+/// Used to describe types defined as structures (e.g., structs in
+/// C/C++).
+TYPE_KIND(StructTyKind)
```
Comment (Snowy1803) [inline include/AST/Types/Typekind.def:4]: You'd put StructTy here, the name of the type
You use NAME ## Kind in your enum macro to add the kind suffix

File: include/AST/Types/TypeBase.hpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
+#ifndef GLU_AST_TYPES_TYPEBASE_HPP_
+#define GLU_AST_TYPES_TYPEBASE_HPP_
+
+#include "llvm/Support/Casting.h"
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:4]: Should be <>

**Review by Snowy1803 — COMMENTED** [review]
Please rebase and clang-format this file

File: include/AST/Types/TypeBase.hpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+///        elements each types should at least have.
+class TypeBase {
+
+public:
+    /// @brief Discriminator for LLVM-style RTTI (used in dyn_cast<> and similar
+    /// operations)
+    enum class TypeKind {
+#define TYPE(NAME) NAME##Kind,
+#include "Types/TypeKind.def"
+#undef TYPE
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:19]: no need to undef, it's already done at the end of the .def

File: include/AST/Types/TypeBase.hpp [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    };
+
+    /// @brief Getter for the kind of the Type.
+    /// @return Returns the size of the type as a TypeKind.
+    TypeKind getKind() const
+    {
+        return _kind;
+    }
```
Comment (Snowy1803) [inline include/AST/Types/TypeBase.hpp:27]: this file is still not clang-formatted

File: include/AST/Types/TypeBase.hpp (hunk 3 above)
Comment (Snowy1803) [reply include/AST/Types/TypeBase.hpp:39]: The enum shouldn't be inside the class, it should be defined before, and outside, in the namespace

File: include/AST/Types/Typekind.def [hunk 10]
```diff
@@ -0,0 +1,67 @@
+#ifndef TYPE
```
Comment (Snowy1803) [inline include/AST/Types/Typekind.def:1]: filename should be TypeKind.def

**Review by Snowy1803 — APPROVED** [review]

**Review by LindonAliu — APPROVED** [review]

**Review by yunsecode — APPROVED** [review]

**Pull request merged**